feat(tagging): add frontend tagging on create new link form

halfwhole opened the pull request:
## Problem

Users should be able to add tags on the create link form

Notion issue [here](https://www.notion.so/opengov/As-a-user-I-want-to-add-tags-while-creating-link-s-so-that-I-can-identify-the-links-easily-7d9b280c038043b9b2210b8eba045fff)

## Solution

Added

These changes are purely on the frontend.

Pending questions:
- Restrictions on 

**Features**:

- Details ...

**Improvements**:

- Details ...

**Bug Fixes**:

- Details ...

## Before & After Screenshots

**BEFORE**:
[insert screenshot here]

**AFTER**:
[insert screenshot here]

## Tests

_What tests should be run to confirm functionality?_

## Deploy Notes

_Notes regarding deployment of the contained body of work. These should note any
new dependencies, new scripts, etc._

**New environment variables**:

- `env var` : env var details

**New scripts**:

- `script` : script details

**New dependencies**:

- `dependency` : dependency details

**New dev dependencies**:

- `dependency` : dependency details

jimvae commented on src/client/user/components/CreateUrlModal/CreateLinkForm.tsx:
```diff
@@ -300,6 +307,76 @@ const CreateLinkForm: FunctionComponent<CreateLinkFormProps> = ({
+            </Typography>
+          </div>
+          <div>
+            <Autocomplete
```
Might want to check these warnings while filling up the Autocomplete text field:

![image](https://user-images.githubusercontent.com/65295129/189035732-d28acd26-4b55-42bb-8d6e-50d33faee94c.png)

halfwhole replied:
thanks for the catch, i'm adding [`freeSolo`](https://v4.mui.com/components/autocomplete/#free-solo) to Autocomplete to stop this from happening!

jimvae commented on src/client/user/components/CreateUrlModal/CreateLinkForm.tsx:
```diff
@@ -300,6 +307,76 @@ const CreateLinkForm: FunctionComponent<CreateLinkFormProps> = ({
+                ))
+              }
+              renderInput={(params) => (
+                <TextField
```
nit: Would it be better to abstract the `Autocomplete` component as another class. It might be cleaner to do so especially when it's not using any shared functions inside the `CreateLinkForm` component

halfwhole replied:
sounds good - i'll extract it as `TagAutocomplete`, and shift it to `client/user/widgets` so that it can be reused by the edit link drawer in the future as well!

jimvae reviewed (approved):
LGTM! The AutoComplete component from mui-lab might have some quirks but it's working fine otherwise.

halfwhole commented on src/client/user/components/CreateUrlModal/FormTag.tsx:
```diff
@@ -0,0 +1,57 @@
+  return (
+    <div className={classes.tag}>
+      <Typography className={classes.tagText}>
+        {tag}{' '}
```
it's unnecessary, my mistake i'll remove this

halfwhole commented on src/client/user/components/CreateUrlModal/CreateLinkForm.tsx:
```diff
@@ -300,6 +307,76 @@ const CreateLinkForm: FunctionComponent<CreateLinkFormProps> = ({
+          <div>
+            <Autocomplete
+              multiple
+              options={['']}
+              value={tags}
+              PaperComponent={React.forwardRef(() => null)}
+              popupIcon={null}
+              closeIcon={null}
```
you're right, i'll remove these components from autocomplete

halfwhole commented on package.json:
```diff
@@ -73,7 +74,7 @@
     "minimatch": "^3.0.4",
     "moment": "^2.29.4",
-    "moment-timezone": "^0.5.33",
+    "moment-timezone": "^0.5.35",
```
yeah strange, commit https://github.com/opengovsg/GoGovSG/commit/2a7e1f2b5b01c9f6bf7f0086287ea403c43650be is already in `feat/tags` 🤔

halfwhole commented:
I've also added checks to disable link creation if the tags are invalid (forgot to add this previously), and refactored `TEXT_FIELD_HEIGHT` to reside under `client/user/constants`

thanhdatle commented on src/shared/util/validation.ts:
```diff
@@ -7,6 +7,10 @@ export const WHITELIST = [new RegExp('^http://localhost:4566')]
 
 export const SHORT_URL_REGEX = /^[a-z0-9-]+$/
 
+export const TAG_STRING_REGEX = /^[A-Za-z0-9-]+$/
```
we need to support underscore as well

halfwhole replied:
checking our [notion doc](https://www.notion.so/opengov/As-a-user-I-want-to-add-tags-while-creating-link-s-so-that-I-can-identify-the-links-easily-7d9b280c038043b9b2210b8eba045fff), it looks like we're not looking to support special characters apart from hyphens! I also modelled this validation on the corresponding section in [this PR](https://github.com/opengovsg/GoGovSG/pull/1917/files#diff-691549b01fd4fb725dccc9297b71b9bdbcee326f2fbe70acc015c235d42a63a8)

halfwhole replied:
(update: we're putting underscores in - taking in the new tag regexes after merging in new changes from `develop`!)

jimvae reviewed (approved):
LGTM!
